sensor/mcp9808/mcp9808.go

```diff
@@ -0,0 +1,484 @@
+// Package mcp9808 is a driver for the MCP9808 temperature sensor
+// and all code is based off of the documentation found here:
+// http://ww1.microchip.com/downloads/en/DeviceDoc/25095A.pdf
+package mcp9808
+
+import (
+	"log"
+	"sync"
+	"time"
+
+	"github.com/kidoman/embd"
+)
+
+const (
+	// default I2C address for device
+	address = 0x18
+
+	// Register addresses.
+	regConfig       = 0x01
+	regUpperTemp    = 0x02
+	regLowerTemp    = 0x03
+	regCriticalTemp = 0x04
+	regAmbientTemp  = 0x05
+	regManufID      = 0x06
+	regDeviceID     = 0x07
+	regResolution   = 0x08
+)
+
+const (
+	configAlertMode uint16 = 1 << iota
+	configAlertPolarity
+	configAlertSelect
+	configAlertControl
+	configAlertStatus
+	configInterruptClear
+	configWindowTempLock
+	configCriticalTempLock
+	configShutDown
+)
+
+// MCP9808 represents a MCP9808 temperature sensor.
+type MCP9808 struct {
+	Bus  embd.I2CBus
+	Poll time.Duration
+
+	cmu sync.RWMutex
```

Comment:
I think the best way to handle the locking is a single simple sync.Mutex. Most or all exported (capitalized) functions start by acquiring the mutex and then deferring an unlock. Few or none of the internal functions touch the mutex. This means that in general an exported function cannot call another exported function 'cause you get a deadlock trying to acquire the mutex twice. In such as situation, you split the called function in two: an internal function with the guts and an exported one that is just a wrapper to acquire the mutex. This design pattern is pretty common in Go in general.